zebra-state/src/service.rs

```diff
@@ -679,3 +690,46 @@ impl Service<Request> for StateService {
 pub fn init(config: Config, network: Network) -> BoxService<Request, Response, BoxError> {
     BoxService::new(StateService::new(config, network))
 }
+
+/// Check if zebra is following a legacy chain and return an error if so.
+pub fn legacy_chain_check<I>(
+    tip_height: block::Height,
+    ancestors: I,
+    network: Network,
+) -> Result<(), BoxError>
+where
+    I: Iterator<Item = Arc<Block>>,
+{
+    const MAX_BLOCKS_TO_CHECK: usize = 100;
+
+    // Only do the check if we have an Nu5 activation height.
+    if let Some(nu5_height) = NetworkUpgrade::Nu5.activation_height(network) {
+        // And only do the check if we are above Nu5 activation.
+        if tip_height >= nu5_height {
+            for (count, block) in ancestors.enumerate() {
+                if count >= MAX_BLOCKS_TO_CHECK {
+                    return Err("giving up after checking too many blocks".into());
+                }
+
+                // All transaction `network_upgrade` fields must be consistent with the block height.
+                if block
+                    .check_transaction_network_upgrade_consistency(network)
+                    .is_err()
+                {
+                    return Err("inconsistent network upgrade found in transaction".into());
+                }
+
+                // If we find at least one transaction with a `network_upgrade` field we are ok.
+                let network_upgrades: Vec<NetworkUpgrade> = block
+                    .transactions
+                    .iter()
+                    .filter_map(|trans| trans.network_upgrade())
+                    .collect();
+                if !network_upgrades.is_empty() {
+                    return Ok(());
+                }
```

Comment:
Minor idea: you can avoid the allocation inside `Vec` and also stop sooner if you use `.find()` or `.find_map()`:

```suggestion
                let has_network_upgrade = block
                    .transactions
                    .iter()
                    .find_map(|trans| trans.network_upgrade())
                    .is_some();
                if has_network_upgrade {
                    return Ok(());
                }
```

Reply:
Nice one, i fixed this at https://github.com/ZcashFoundation/zebra/pull/2366/commits/4c6e9379f4b837e4e1989038fa0b942a74789443